---
title: "Example of label projection"
output: html_notebook
---

# Acquire the eigenProjectR pre-made list objects
From [INSERT IN URL]
```{r}
load('~/Downloads/scEiaD_seacell_pca_objs.Rdata')
# load('inst/data/scEiaD_seacell_pca_objs.Rdata')
```

# Acquire outside data
https://singlecell.broadinstitute.org/single_cell/study/SCP1755/joint-scrna-seq-and-scatac-seq-atlas-of-the-adult-human-retina#study-download

Download:
  - processed_barcodes.tsv.gz 
  - processed_features.tsv.gz 
  - processed_matrix.mtx.gz 
  
Make folder "SCP1755"

Put those three files into SCP1755 and hand delete the "processed_" part from each of the three names.

Also Download:
  - scPortal_metadata_file.txt 

# Example of how SCP755 should look (I did this all in my ~/Downloads folder)
```{r}
list.files('~/Downloads/SCP1755')
```


# Load in SCP755 
After loading the counts in as a `SingleCellExperiment` object, we then add in the cell type assignments to each cell
```{r}
library(tidyverse)
library(scran)
library(eigenProjectR)

scp1755 <- DropletUtils::read10xCounts('~/Downloads/SCP1755/')
scp1755_meta <- read_tsv('~/Downloads/scPortal_metadata_file.txt')

scp1755_meta <- scp1755_meta[2:nrow(scp1755_meta),]

colData(scp1755)$NAME <- colnames(assay(scp1755))

colData(scp1755)$celltype <- colData(scp1755) %>% as_tibble() %>% mutate(NAME = Barcode) %>% left_join(scp1755_meta) %>% pull(cell_type__custom)
```

# Aggregate to the cell type level
```{r}
scp1755_sum_counts <- aggregateAcrossCells(scp1755, colData(scp1755)$celltype) %>% assay()
row.names(scp1755_sum_counts) <- rowData(scp1755)$Symbol
```

# PROFIT!!!!!!!!!!!
```{r}
scp1755_projected <- eigenProjectR(scp1755_sum_counts, rotation = seacell_pca_human_adult$PCA$rotation)


lr_model <- model_build(seacell_pca_human_adult$PCA$x[,1:250],
                        seacell_pca_human_adult$meta$CellType,
                        BPPARAM = BiocParallel::MulticoreParam(4), verbose = FALSE)

model_apply(lr_model, scp1755_projected)
```


##############################
#
##############################



##############################

PCA <- scp1755_project$existing_data_projected_objects$PCA
dataGG <- scp1755_project$existing_data_projected_objects$meta
percentVar <- scp1755_project[[3]][[3]]
select <- scp1755_project[[3]][[4]]

pcFirst <- 'PC1'
pcSecond <- 'PC2'
rotations <- c(pcFirst, pcSecond)


top_rotations <-
  c(PCA$rotation[,str_extract(pcFirst, '\\d+') %>% as.integer()] %>% sort() %>% head(3) %>% names(),
    PCA$rotation[,str_extract(pcFirst, '\\d+') %>% as.integer()] %>% sort() %>% tail(3) %>% names(),
    PCA$rotation[,str_extract(pcSecond, '\\d+') %>% as.integer()] %>% sort() %>% head(3) %>% names(),
    PCA$rotation[,str_extract(pcSecond, '\\d+') %>% as.integer()] %>% sort() %>% tail(3) %>% names()) %>%
  unique()

rotation_multipler_first <- dataGG[pcFirst] %>% pull(1) %>% abs() %>% max() / PCA$rotation[,str_extract(pcFirst, '\\d+') %>% as.integer()] %>% abs() %>% max()
rotation_multipler_second <- dataGG[pcSecond] %>% pull(1) %>% abs() %>% max() / PCA$rotation[,str_extract(pcSecond, '\\d+') %>% as.integer()] %>% abs() %>% max()

p <- bind_rows(dataGG %>% as_tibble() %>% mutate(label = paste0(study_accession, '\n',CTP)),
               scp1755_project[[1]] %>% as_tibble(rownames = 'label') %>% mutate(CellType = 'scp1755')) %>%
  as_tibble() %>%
  ggplot(., aes(.data[[pcFirst]], .data[[pcSecond]])) +
  geom_point(size=0.5, aes(color=CellType,
                           label = label)) +
  geom_segment(data = PCA$rotation[top_rotations,rotations] %>% data.frame(), aes(x=0,y=0, xend = .data[[pcFirst]]*rotation_multipler_first*0.95, yend = .data[[pcSecond]]*rotation_multipler_second*0.95)) +
  geom_text(data = PCA$rotation[top_rotations,rotations] %>%
              as_tibble(rownames = 'Gene') %>%
              mutate(Gene = gsub(' \\(.*','',Gene)),
            aes(x=.data[[pcFirst]]*rotation_multipler_first*1.0,
                y = .data[[pcSecond]] * rotation_multipler_second*1.0,
                label = Gene)) +
  xlab(paste0(pcFirst, ": ",percentVar[str_extract(pcFirst, '\\d+') %>% as.integer()],"% variance")) +
  ylab(paste0(pcSecond, ": ",percentVar[str_extract(pcSecond, '\\d+') %>% as.integer()],"% variance")) +
  cowplot::theme_cowplot() +
  scale_color_manual(values = c(pals::glasbey(), pals::alphabet2(), pals::alphabet2()) %>% unname()) +
  scale_fill_manual(values = c(pals::glasbey(), pals::alphabet2(), pals::alphabet2()) %>% unname()) +
  scale_shape_manual(values = 0:10)
plotly::ggplotly(p)


```
